Support for additional filters/jails added

apazga opened the pull request:
I added support for including additional filters and added some common jails people use (and possibly more coming soon).

Due to new jails need additional filters, I added an optional parameter "additional_jails" (default: false) that deploys the additional filters needed for them.

Adding new filters from now on will be easy too, due to all filters included in files folder will be synchronized.

jg-development commented:
+1 for merging

jg-development commented:
How about a define like the "fail2ban::define"?
With a "fail2ban::filter" with hiera hash it could be possible to add filters dynamic.

apazga commented:
Finally had some time (and also the need!) to improve this.
@jg-development I added that requested define type for creating new jails (and filters) and updated all doc & examples.

dhoppe commented:
@apazga I like the idea of managing the filter and jails in an global directory for alle sorts of operating systems. Could you please do a rebase against the latest changes at master?

apazga commented:
Done! :bowtie:

dhoppe commented:
@apazga Now you need to fix the RSpec tests, because they do not know anything about the `$::osfamily` directory. 

I also think you should move the `$::lsbdistcodename`specific directories into the new `$::osfamily` namespace.

![screen shot 2016-11-26 at 08 18 46](https://cloud.githubusercontent.com/assets/1111056/20638809/7c2bd6e8-b3b1-11e6-9285-4e762da6055a.png)

dhoppe commented on README.md:
````diff
@@ -65,13 +65,13 @@ Install fail2ban with the recommended parameters.
 ```puppet
   class { 'fail2ban':
-    config_file_template => "fail2ban/%{::osfamily}/etc/fail2ban/jail.conf.erb"
+    config_file_template => "fail2ban/RedHat/Carbon/etc/fail2ban/jail.conf.erb"
````
Please use `%{::osfamily}` and `%{::lsbdistcodename}` in case someone wants to copy & paste from the documentation.

dhoppe commented on README.md:
```diff
@@ -147,15 +168,15 @@ Deploy additional configuration files from source, string or template.
         'jail.2nd.conf' => {
           config_file_path   => '/etc/fail2ban/jail.2nd.conf',
-          config_file_source => "puppet:///modules/fail2ban/${::lsbdistcodename}/etc/fail2ban/jail.2nd.conf",
+          config_file_source => "puppet:///modules/fail2ban/${::osfamily}/${::lsbdistcodename}/etc/fail2ban/jail.2nd.conf",
```
Please remove the line break at line 171.

dhoppe commented on README.md:
```diff
@@ -313,13 +334,155 @@ Defaults to '3'.
+
+Determines which custom jails should be included (see [Custom jails](#custom-jails).
+
+## Jails available
```
Please add a newline after each heading. This is required by Markdown lint.

dhoppe commented on README.md:
```diff
@@ -313,13 +334,155 @@ Defaults to '3'.
 * Debian 6/7/8
 * Ubuntu 12.04/14.04/16.04
 * RedHat 5/6/7
+* CentOS 6/7
```
Are these releases covered at `metadata.json`?

dhoppe commented:
The RSpec tests still does not know anything about the new `${::osfamily}` based directories. For example `fail2ban/wheezy/etc/fail2ban/jail.conf.erb` vs `fail2ban/Debian/wheezy/etc/fail2ban/jail.conf.erb`.

apazga commented:
Done! I think it's ok now.

dhoppe commented:
@apazga Thx for your contribution / patience. This is a big improvement.

apazga commented:
You're welcome. I'm glad to help. I'm also using this module in my company, so thank you too :thumbsup:

Kriechi commented:
This PR just introduced an API-incompatible change with a patch-level version bump.
The path was changed: `modules/fail2ban/${::osfamily}/${::lsbdistcodename}`
Not very nice in terms of semver... 👎

dhoppe commented:
@Kriechi I had a feeling that this might happen and removed the `${::osfamily}` part.

Kriechi commented:
Thanks for reverting this - but now you violated semver twice 😉 
Still - I think it is better this way for most other people who use this module.
